### ISSUE TITLE
More featureful convolutions

### ISSUE THREAD

**Issue by Giels** [arrayfire/arrayfire#1719]
The convolution implementation is currently limited.

Firstly, the only operating modes involve padding the input with 0's before convolution, to reach an output size which is the same as the input's, or padding the input with 0's before convolution, to reach an output size larger than the input's.
However, it is often desirable to operate in different ways: input wrap, input mirror, input stretch or discard are all standard modes in many convolution implementations.
Alternatively, the user could be in charge of tiling and the convolution should discard ('valid' convolution in matlab) in default mode (still expanding in expand mode).

Secondly, it does not support strides, a common, and very useful, feature in most convolution implementations.

**Comment by kkotyk** [comment]
By supporting strides, do you mean to incorporate decimation? Because if so that would be invaluable to my work and that of others. I'm not a developer for this package, but I just thought I would support this idea.

**Comment by pavanky** [comment]
@kkotyk I haven't heard the term decimation, but strided convolutions are used a lot in machine learning. Can you point me to some docs about this ?

**Comment by kkotyk** [comment]
@pavanky Perhaps this strided convolution refers to something else. I'm not that familiar with that term. Decimation and interpolation are techniques used to change the sampling rate of a signal. Here is a decent PDF on it. https://www.staff.ncl.ac.uk/oliver.hinton/eee305/Chapter9.pdf. In decimation for example, one cannot simply downsample a digital signal to achieve a lower sampling rate. You have to lowpass the signal first in order to eliminate aliasing and to filter first you do convolution. But because you are downsampling as well, an optimization can be applied. If you are downsampling by a rate, N, then the convolution calculation only needs to be performed for every N samples and the rest can be discarded. It's pretty much the opposite process for interpolation. Maybe something exists like this in arrayfire and I just don't know about it, but having optimized multirate techniques would be valuable for realtime DSP applications.

Page 4 of this PDF shows another technique than can be used for decimation, but I mainly link this because the diagram shows how the convolution can be skipped for the samples being discarded.
http://www.ws.binghamton.edu/fowler/fowler%20personal%20page/EE521_files/IV-05%20Polyphase%20FIlters_2007.pdf

I'm a pretty capable developer and could contribute to something like this, but I would have to work with someone who is more knowledgeable on GPU algorithms.

**Comment by pavanky** [comment]
@kkotyk it sort of sounds like a strided convolution (or they atleast capture a part of the entire process).

CC'ing @kaatish.

**Comment by pavanky** [comment]
@kkotyk I'll read the paper to understand it a bit more :)

**Comment by kkotyk** [comment]
@pavanky Excellent. I am interested to hear your thoughts.

**Comment by Giels** [comment]
@kkotyk I was thinking of strided convolution in the sense of machine learning, as @pavanky mentions. Equation 9.1 in your first link precisely describes the operation in functional terms, so it should be synonymous to decimation.

**Comment by kkotyk** [comment]
@Giels Right, so it appears this would be a feature that is useful in several different applications. Thanks for taking a look.

**Comment by Giels** [comment]
Any update on this? @pavanky ?

**Comment by Thormidable** [comment]
Could I put in my support for these additional features? Particularly variants of padding.

**Comment by Thormidable** [comment]
Can I add my support for this feature request?

**Comment by kkotyk** [comment]
@Thormidable Are you able to create a pull request so the project members can review your work?

**Comment by kkotyk** [comment]
@Thormidable Or are you just in favor for this too :p

**Comment by Thormidable** [comment]
I'm just in favour of this. 

I have (very quickly) bashed together some very simple arrayfire code to do padding of various types. It's pretty simple  but I'm happy to drop it here if it would be useful.
